### ISSUE TITLE
cascade vs list : device_infos discrepancy

### ISSUE THREAD

**Issue by MichaelLiss** [jimmykane/easy-fit#18]
Hi again,

Please forgive my ignorance if this is supposed to behave this way, but I am comparing the output from the 'cascade' option and the 'list' option and noticed a discrepancy between the two.  The 'list' version has multiple device_info records and the cascade does not have any?

![Cascade_vs_List](https://user-images.githubusercontent.com/39492750/80821124-0512f700-8b8d-11ea-9b08-48a0f4510cbc.png)

Here is the full output from both:

**Cascade**

```javascript
{
  "file_id": {
    "device_id": 3996694396,
    "time_created": "2020-04-30T22:16:48.000Z",
    "manufacturer": "garmin",
    "product": 3113,
    "type": "activity"
  },
  "file_creator": {
    "software_version": 440
  },
  "device_settings": {
    "utc_offset": 0,
    "time_offset": 4294942096,
    "active_time_zone": 0,
    "time_zone_offset": 0,
    "mounting_side": "left"
  },
  "user_profile": {
    "weight": 74.8,
    "gender": "male",
    "height": 0.00183,
    "language": "english",
    "elev_setting": "statute",
    "weight_setting": "statute",
    "resting_heart_rate": 0,
    "hr_setting": "max",
    "speed_setting": "statute",
    "dist_setting": "statute",
    "position_setting": "degree_minute_second",
    "temperature_setting": "statute",
    "height_setting": "statute"
  },
  "sport": {
    "name": "Run",
    "sport": "running",
    "sub_sport": "generic"
  },
  "zones_target": {
    "functional_threshold_power": 207,
    "max_heart_rate": 180,
    "threshold_heart_rate": 167,
    "pwr_calc_type": "percent_ftp"
  },
  "activity": {
    "timestamp": "2020-04-30T22:17:03.000Z",
    "total_timer_time": 9.199,
    "local_timestamp": "2020-04-30T15:17:03.000Z",
    "num_sessions": 1,
    "type": "manual",
    "event": "activity",
    "event_type": "stop",
    "sessions": [
      {
        "timestamp": "2020-04-30T22:17:03.000Z",
        "start_time": "2020-04-30T22:16:48.000Z",
        "start_position_lat": 37.711845925077796,
        "start_position_long": -121.72617522999644,
        "total_elapsed_time": 9.199,
        "total_timer_time": 9.199,
        "total_distance": 0,
        "nec_lat": 37.711845925077796,
        "nec_long": -121.726168775931,
        "swc_lat": 37.711771074682474,
        "swc_long": -121.72637580893934,
        "enhanced_avg_speed": 0,
        "enhanced_max_speed": 0,
        "message_index": {
          "0": false,
          "value": 0,
          "reserved": false,
          "selected": false
        },
        "total_calories": 0,
        "total_ascent": 0,
        "total_descent": 0,
        "first_lap_index": 0,
        "num_laps": 1,
        "left_right_balance": {
          "0": false,
          "value": 16383,
          "right": true
        },
        "event": "lap",
        "event_type": "stop",
        "sport": "running",
        "sub_sport": "generic",
        "trigger": "activity_end",
        "avg_temperature": -246.14999999999998,
        "max_temperature": -246.14999999999998,
        "laps": [
          {
            "timestamp": "2020-04-30T22:17:03.000Z",
            "start_time": "2020-04-30T22:16:48.000Z",
            "start_position_lat": 37.711845925077796,
            "start_position_long": -121.72617522999644,
            "end_position_lat": 37.711771074682474,
            "end_position_long": -121.72633968293667,
            "total_elapsed_time": 9.199,
            "total_timer_time": 9.199,
            "total_distance": 0,
            "total_cycles": 0,
            "enhanced_avg_speed": 0,
            "enhanced_max_speed": 0,
            "message_index": {
              "0": false,
              "value": 0,
              "reserved": false,
              "selected": false
            },
            "total_calories": 0,
            "total_ascent": 0,
            "total_descent": 0,
            "left_right_balance": {
              "0": false,
              "value": 16383,
              "right": true
            },
            "wkt_step_index": {
              "0": false,
              "value": 4095,
              "reserved": true,
              "selected": true
            },
            "event": "lap",
            "event_type": "stop",
            "avg_cadence": 0,
            "max_cadence": 0,
            "lap_trigger": "session_end",
            "sport": "running",
            "sub_sport": "generic",
            "avg_temperature": -246.14999999999998,
            "max_temperature": -246.14999999999998,
            "avg_fractional_cadence": 0,
            "max_fractional_cadence": 0,
            "records": [
              {
                "timestamp": "2020-04-30T22:16:48.000Z",
                "elapsed_time": 0,
                "timer_time": 0,
                "position_lat": 37.711845925077796,
                "position_long": -121.72617522999644,
                "distance": 0,
                "enhanced_speed": 0,
                "enhanced_altitude": 0.1624,
                "cadence": 0,
                "temperature": -246.14999999999998,
                "fractional_cadence": 0
              },
              {
                "timestamp": "2020-04-30T22:16:54.000Z",
                "elapsed_time": 6,
                "timer_time": 6,
                "position_lat": 37.71178691647947,
                "position_long": -121.72637220472097,
                "distance": 0,
                "enhanced_speed": 0,
                "enhanced_altitude": 0.163,
                "cadence": 0,
                "temperature": -246.14999999999998,
                "fractional_cadence": 0
              },
              {
                "timestamp": "2020-04-30T22:16:58.000Z",
                "elapsed_time": 10,
                "timer_time": 10,
                "position_lat": 37.711771074682474,
                "position_long": -121.72633968293667,
                "distance": 0,
                "enhanced_speed": 0,
                "enhanced_altitude": 0.163,
                "cadence": 0,
                "temperature": -246.14999999999998,
                "fractional_cadence": 0
              }
            ]
          }
        ]
      }
    ],
    "events": [
      {
        "timestamp": "2020-04-30T22:16:48.000Z",
        "data": 0,
        "event": "timer",
        "event_type": "start",
        "event_group": 0
      },
      {
        "timestamp": "2020-04-30T22:16:58.000Z",
        "data": 0,
        "event": "timer",
        "event_type": "stop_all",
        "event_group": 0
      },
      {
        "timestamp": "2020-04-30T22:17:02.000Z",
        "data": 207,
        "event_type": "marker",
        "event_group": 1
      }
    ],
    "hrv": []
  }
}
```

**List**

```javascript
{
  "file_id": {
    "device_id": 3996694396,
    "time_created": "2020-04-30T22:16:48.000Z",
    "manufacturer": "garmin",
    "product": 3113,
    "type": "activity"
  },
  "file_creator": {
    "software_version": 440
  },
  "device_settings": {
    "utc_offset": 0,
    "time_offset": 4294942096,
    "active_time_zone": 0,
    "time_zone_offset": 0,
    "mounting_side": "left"
  },
  "user_profile": {
    "weight": 74.8,
    "gender": "male",
    "height": 0.00183,
    "language": "english",
    "elev_setting": "statute",
    "weight_setting": "statute",
    "resting_heart_rate": 0,
    "hr_setting": "max",
    "speed_setting": "statute",
    "dist_setting": "statute",
    "position_setting": "degree_minute_second",
    "temperature_setting": "statute",
    "height_setting": "statute"
  },
  "sport": {
    "name": "Run",
    "sport": "running",
    "sub_sport": "generic"
  },
  "zones_target": {
    "functional_threshold_power": 207,
    "max_heart_rate": 180,
    "threshold_heart_rate": 167,
    "pwr_calc_type": "percent_ftp"
  },
  "activity": {
    "timestamp": "2020-04-30T22:17:03.000Z",
    "total_timer_time": 9.199,
    "local_timestamp": "2020-04-30T15:17:03.000Z",
    "num_sessions": 1,
    "type": "manual",
    "event": "activity",
    "event_type": "stop"
  },
  "sessions": [
    {
      "timestamp": "2020-04-30T22:17:03.000Z",
      "start_time": "2020-04-30T22:16:48.000Z",
      "start_position_lat": 37.711845925077796,
      "start_position_long": -121.72617522999644,
      "total_elapsed_time": 9.199,
      "total_timer_time": 9.199,
      "total_distance": 0,
      "nec_lat": 37.711845925077796,
      "nec_long": -121.726168775931,
      "swc_lat": 37.711771074682474,
      "swc_long": -121.72637580893934,
      "enhanced_avg_speed": 0,
      "enhanced_max_speed": 0,
      "message_index": {
        "0": false,
        "value": 0,
        "reserved": false,
        "selected": false
      },
      "total_calories": 0,
      "total_ascent": 0,
      "total_descent": 0,
      "first_lap_index": 0,
      "num_laps": 1,
      "left_right_balance": {
        "0": false,
        "value": 16383,
        "right": true
      },
      "event": "lap",
      "event_type": "stop",
      "sport": "running",
      "sub_sport": "generic",
      "trigger": "activity_end",
      "avg_temperature": -246.14999999999998,
      "max_temperature": -246.14999999999998
    }
  ],
  "laps": [
    {
      "timestamp": "2020-04-30T22:17:03.000Z",
      "start_time": "2020-04-30T22:16:48.000Z",
      "start_position_lat": 37.711845925077796,
      "start_position_long": -121.72617522999644,
      "end_position_lat": 37.711771074682474,
      "end_position_long": -121.72633968293667,
      "total_elapsed_time": 9.199,
      "total_timer_time": 9.199,
      "total_distance": 0,
      "total_cycles": 0,
      "enhanced_avg_speed": 0,
      "enhanced_max_speed": 0,
      "message_index": {
        "0": false,
        "value": 0,
        "reserved": false,
        "selected": false
      },
      "total_calories": 0,
      "total_ascent": 0,
      "total_descent": 0,
      "left_right_balance": {
        "0": false,
        "value": 16383,
        "right": true
      },
      "wkt_step_index": {
        "0": false,
        "value": 4095,
        "reserved": true,
        "selected": true
      },
      "event": "lap",
      "event_type": "stop",
      "avg_cadence": 0,
      "max_cadence": 0,
      "lap_trigger": "session_end",
      "sport": "running",
      "sub_sport": "generic",
      "avg_temperature": -246.14999999999998,
      "max_temperature": -246.14999999999998,
      "avg_fractional_cadence": 0,
      "max_fractional_cadence": 0
    }
  ],
  "records": [
    {
      "timestamp": "2020-04-30T22:16:48.000Z",
      "elapsed_time": 0,
      "timer_time": 0,
      "position_lat": 37.711845925077796,
      "position_long": -121.72617522999644,
      "distance": 0,
      "enhanced_speed": 0,
      "enhanced_altitude": 0.1624,
      "cadence": 0,
      "temperature": -246.14999999999998,
      "fractional_cadence": 0
    },
    {
      "timestamp": "2020-04-30T22:16:54.000Z",
      "elapsed_time": 6,
      "timer_time": 6,
      "position_lat": 37.71178691647947,
      "position_long": -121.72637220472097,
      "distance": 0,
      "enhanced_speed": 0,
      "enhanced_altitude": 0.163,
      "cadence": 0,
      "temperature": -246.14999999999998,
      "fractional_cadence": 0
    },
    {
      "timestamp": "2020-04-30T22:16:58.000Z",
      "elapsed_time": 10,
      "timer_time": 10,
      "position_lat": 37.711771074682474,
      "position_long": -121.72633968293667,
      "distance": 0,
      "enhanced_speed": 0,
      "enhanced_altitude": 0.163,
      "cadence": 0,
      "temperature": -246.14999999999998,
      "fractional_cadence": 0
    }
  ],
  "events": [
    {
      "timestamp": "2020-04-30T22:16:48.000Z",
      "data": 0,
      "event": "timer",
      "event_type": "start",
      "event_group": 0
    },
    {
      "timestamp": "2020-04-30T22:16:58.000Z",
      "data": 0,
      "event": "timer",
      "event_type": "stop_all",
      "event_group": 0
    },
    {
      "timestamp": "2020-04-30T22:17:02.000Z",
      "data": 207,
      "event_type": "marker",
      "event_group": 1
    }
  ],
  "device_infos": [
    {
      "timestamp": "2020-04-30T22:16:48.000Z",
      "serial_number": 3996694396,
      "manufacturer": "garmin",
      "product": 3113,
      "software_version": 4.4,
      "device_index": 0,
      "source_type": "local"
    },
    {
      "timestamp": "2020-04-30T22:16:48.000Z",
      "manufacturer": "garmin",
      "product": 3113,
      "software_version": 4.4,
      "device_index": 1,
      "source_type": "local"
    },
    {
      "timestamp": "2020-04-30T22:16:48.000Z",
      "manufacturer": "garmin",
      "product": 3107,
      "software_version": 4.1,
      "device_index": 2,
      "device_type": 0,
      "source_type": "local"
    },
    {
      "timestamp": "2020-04-30T22:16:48.000Z",
      "product": 8195,
      "software_version": 0,
      "device_index": 3,
      "source_type": "local"
    },
    {
      "timestamp": "2020-04-30T22:16:48.000Z",
      "manufacturer": 0,
      "product": 0,
      "software_version": 1.28,
      "device_index": 4,
      "source_type": "local"
    },
    {
      "timestamp": "2020-04-30T22:16:48.000Z",
      "manufacturer": "garmin",
      "product": 3114,
      "software_version": 4.01,
      "device_index": 5,
      "device_type": "environment_sensor_legacy",
      "source_type": "local"
    },
    {
      "timestamp": "2020-04-30T22:16:58.000Z",
      "serial_number": 3996694396,
      "manufacturer": "garmin",
      "product": 3113,
      "software_version": 4.4,
      "device_index": 0,
      "source_type": "local"
    },
    {
      "timestamp": "2020-04-30T22:16:58.000Z",
      "manufacturer": "garmin",
      "product": 3113,
      "software_version": 4.4,
      "device_index": 1,
      "source_type": "local"
    },
    {
      "timestamp": "2020-04-30T22:16:58.000Z",
      "manufacturer": "garmin",
      "product": 3107,
      "software_version": 4.1,
      "device_index": 2,
      "device_type": 0,
      "source_type": "local"
    },
    {
      "timestamp": "2020-04-30T22:16:58.000Z",
      "product": 8195,
      "software_version": 0,
      "device_index": 3,
      "source_type": "local"
    },
    {
      "timestamp": "2020-04-30T22:16:58.000Z",
      "product": 28745,
      "software_version": 1.28,
      "device_index": 4,
      "source_type": "local"
    },
    {
      "timestamp": "2020-04-30T22:16:58.000Z",
      "manufacturer": "garmin",
      "product": 3114,
      "software_version": 4.01,
      "device_index": 5,
      "device_type": "environment_sensor_legacy",
      "source_type": "local"
    }
  ],
  "developer_data_ids": [],
  "field_descriptions": [],
  "hrv": [],
  "dive_gases": [],
  "course_points": []
}
```

**Comment by jimmykane** [comment]
Will have to check this

Hold on there. Just need some proper time and as a new dad its hard to find

**Comment by MichaelLiss** [comment]
Congratulations !

and I understand !

Mike

**Comment by MichaelLiss** [comment]
Hello again... 
I was wondering if you have some time to address this ?
I am getting to the point where this issue is starting to block the progress on my project.
(or maybe you could send some instructions on how to modify this so I can update it... and I will try to do the work???)

**Comment by jimmykane** [comment]
I am a bit on vacation but I'll take a look on this one as indeed looks
wierd.

I hope end of weekend.

Don't worry this lib is not forgotten , it's actually used by my main
project, if you want to take a look at it quantified-self.io

Is that ok for you?

**Comment by MichaelLiss** [comment]
Hi...

I am going to uninstall your module and use Garmin's parser and an AWS micro-service, as it doesn't seem you are going to work on this any time soon.

**Comment by jimmykane** [comment]
Hey ! I am back! Should I go on fixing this ?

**Comment by MichaelLiss** [comment]
yes

**Comment by jimmykane** [comment]
Should be fixed with 1.7.0
